### PR TITLE
[Migration Engine Part 1] Implement the foundation for migration engine

### PR DESCRIPTION
**Pull request by warwickschroeder**
## What this adds

The foundation for moving an error instance's data from RavenDB to SQL Server or PostgreSQL. **Nothing is copied yet**, and the migration is not switched on anywhere.

- **A read-only RavenDB source.** `RavenMigrationSource` opens both RavenDB databases, embedded or external, through a client that refuses anything outside a read allow-list (`OnBeforeRequest`, non-tracking sessions), so a patch, a bulk insert or a HiLo range throws in-process rather than reaching the database. It can describe itself and count every collection; `Count`, `Read` and `ReadBody` for a category all throw `NotSupportedException`, so no reader exists yet.
- **`--migration-source-report`.** Prints the source's RavenDB version, where it is, both database names with the setting each came from, and a row count per collection. It reads the instance's own RavenDB settings; `ServiceControl/Migration/SourcePersistenceType` defaults to `RavenDB`, and `Settings.MigrationSourcePersistenceType` now appears in the approved platform sample settings.
- **Two persisters in one process.** `PersistenceFactory.CreateMigrationSource` loads the source persister in its own `AssemblyLoadContext`, beside the configured one, and `OpenMigrationSource` opens it.
- **The migration engine, against fakes.** Store-neutral seams in `ServiceControl.Persistence/DataMigration` (`IMigrationSource`, `IMigrationTarget`, `IMigrationCheckpointStore`, the category registry) and `MigrationEngine`, which copies a category in batches with checkpoint resume, ordering between categories, body read retries, throttling, skip reasons and a halt threshold. A checkpoint carries a `Version` and the target extends it inside the transaction that writes the rows, so a save that loses a race throws `MigrationCheckpointConflictException` and leaves the other writer's row standing rather than halting the category. There is no SQL target yet; it runs against in-memory fakes.
- **Docs and CI.** `docs/migration/` holds the overview, a system design diagram, and short instructions for what can be run today, linked from the README. `.github/workflows/ci.yml` gains a `Migration` test category, which also downloads the RavenDB server.

## Tests

- `ServiceControl.UnitTests/Migration`: engine, registry, options and the fakes.
- `ServiceControl.UnitTests/Hosting`: `--migration-source-report` argument parsing.
- `ServiceControl.UnitTests/ApprovalFiles`: the platform sample settings, which gained the new setting.
- `ServiceControl.Persistence.Tests.RavenDB/DataMigration`: the read-only source lifecycle.
- `ServiceControl.Migration.Tests` (new): the source report command, and both persisters loaded in one process.

### CODE REVIEW

File: docs/migration/ravendb-to-sql-migration-instructions.md [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+| Setting | Environment variable | What it is |
+| --- | --- | --- |
+| `ServiceControl/RavenDB/ConnectionString` | `SERVICECONTROL_RAVENDB_CONNECTIONSTRING` | An external RavenDB server. Leave unset for an embedded database |
+| `ServiceControl/DbPath` | `SERVICECONTROL_DBPATH` | The embedded database's data directory |
+| `ServiceControl/RavenDB/DatabaseName` | `SERVICECONTROL_RAVENDB_DATABASENAME` | The primary database, `primary` by default |
+| `LicensingComponent/RavenDB/ThroughputDatabaseName` | `LICENSINGCOMPONENT_RAVENDB_THROUGHPUTDATABASENAME` | The throughput database, `throughput` by default |
+| `ServiceControl/RavenDB/ClientCertificatePath` or `ServiceControl/RavenDB/ClientCertificateBase64`, with `ServiceControl/RavenDB/ClientCertificatePassword` | `SERVICECONTROL_RAVENDB_CLIENTCERTIFICATEPATH` and so on | A secured external server's client certificate |
+| `ServiceControl/ErrorRetentionPeriod` | `SERVICECONTROL_ERRORRETENTIONPERIOD` | Required. Don't change it during the move |
+
+`ServiceControl/Migration/SourcePersistenceType` defaults to `RavenDB` and needs no setting.
```
Comment (rbev) [inline docs/migration/ravendb-to-sql-migration-instructions.md:21]: So given the abstractions does this technically allow two way migration? or will that be missing the migrationsource implementation?

If this defaults to the only permissible value should it even be a configuration knob?

File: docs/migration/ravendb-to-sql-migration-overview.md [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+- Primary and throughput RavenDB databases in different locations
+- Anything but RavenDB as the source, or anything but a ServiceControl EF Core persister as the target
+
+## Migration workflow
+
+1. Upgrade ServiceControl as normal, still on RavenDB.
+2. Set four things in configuration: the new `PersistenceType`, its connection string, `MigrationMode=true`, and which [optional data](#data-to-be-migrated-categories) they want copied.
+3. Run `--setup` to create the SQL schema. It fails against a SQL Server instance without Full-Text Search installed.
+4. Run the [dry run](#dry-run). It reports what it resolved as a source, what each category holds, and an estimate of how long ServiceControl will be closed. Read [what the dry run reports](#dry-run) before booking an outage around its estimate.
```
Comment (rbev) [inline docs/migration/ravendb-to-sql-migration-overview.md:75]: Should make sure that dry-run checks that the sql schema has been created, just for completeness given item 3

File: docs/migration/ravendb-to-sql-migration-overview.md [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+- `UniqueMessageId` keeps its value, but converts type: the source holds a string and the target column is a `uniqueidentifier`. It is the primary key, the ServicePulse URL, the retry correlation key and the body lookup key at once.
+- `StatusChangedAt` is reconstructed from `@expires` for resolved and archived messages, which is the only place RavenDB sets it. Unresolved and retry-issued messages have no `@expires`, so the copier uses the newest processing attempt's timestamp. The column is `NOT NULL`, so it cannot be left empty, but the value is harmless for those two: the retention sweep only considers resolved and archived rows, so an unresolved message never ages out whatever is written here.
+- Message bodies go through `IBodyStoragePersistence`, which owns the compression threshold and the choice of filesystem, Azure Blob or S3. The separate 102,400-byte inline threshold is not there: it lives on the ingestion path, so the copier has to apply it rather than inherit it.
+- Throughput rows are written directly rather than through the collector, and the write sets each day's count rather than adding to it.
```
Comment (rbev) [inline docs/migration/ravendb-to-sql-migration-overview.md]: Is throughput copied after SC starts up?
Overwrite rather than merge seems problematic here if it is, if not then it needs to be called out

File: docs/migration/ravendb-to-sql-migration-overview.md [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+
+**A category can finish with a small amount of loss and still count as complete.** A few skipped rows in a large table leave the category in a *complete with errors* state, which blocks nothing. Its skipped count is printed and the ids of the skipped rows are written to the log, so while the RavenDB database still exists you can go and look at exactly what did not make it.
+
+## The one point you can go back
+
+While ServiceControl is closed and the required copy is running, nothing except the copier has written to SQL, and the migration has written nothing to RavenDB, which is still authoritative. RavenDB's own expiration still runs, though: unless you disabled it, it keeps deleting expired failed messages and event log items, as [Goals](#goals) describes. Back up both RavenDB databases, or disable expiration on them, before you start. If you need your instance back, set `MigrationMode=false`, point `PersistenceType` back at RavenDB, and start. You lose the copy, not your data, and you can start again later.
```
Comment (rbev) [inline docs/migration/ravendb-to-sql-migration-overview.md]: How does the count work when ravenDB is actively deleting rows from underneath you? Do they just end up in a "skipped" bucket with an unknown reason because you ran off the end of the work with less rows than you expected?

File: src/ServiceControl.Persistence/DataMigration/IMigrationTarget.cs [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+/// <summary>Implemented by a persister that can be the new database a migration copies into.</summary>
+public interface IMigrationTarget
+{
+    /// <summary>How many rows to read per batch for this category. The target picks it because its own database sets the limits.</summary>
+    int BatchSizeFor(MigrationCategory category);
```
Comment (rbev) [inline src/ServiceControl.Persistence/DataMigration/IMigrationTarget.cs:11]: Should this be `MaximumBatchSize` so it's explicit that it's a limit rather than a target to the consumer?

File: src/ServiceControl.Persistence/DataMigration/IMigrationTarget.cs [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+
+    /// <summary>Saves the batch's rows and checkpointAfterBatch in one go, so progress never gets ahead of the data. Save the checkpoint exactly as given, without adding counts to it.</summary>
```
Comment (rbev) [inline src/ServiceControl.Persistence/DataMigration/IMigrationTarget.cs]:
```suggestion
    /// <summary>Saves the batch's rows and checkpointAfterBatch in one transaction, so progress is consistent with the data. Save the checkpoint exactly as given, without adding counts to it.</summary>
```

File: src/ServiceControl.Persistence/DataMigration/MigrationEngine.cs [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+                {
+                    Cursor = batch.Cursor,
+                    CopiedCount = checkpoint.CopiedCount + batchToWrite.Rows.Count,
+                    SkippedCount = checkpoint.SkippedCount + bodySkips,
+                    SkipReasons = AddSkipReasons(checkpoint.SkipReasons, bodySkips == 0 ? null : new Dictionary<string, long> { [nameof(MigrationSkipReason.BodyUnreadable)] = bodySkips })
+                };
+
+                var result = await target.Write(category, batchToWrite, checkpointAfterBatch, cancellationToken);
+
+                checkpoint = checkpointAfterBatch with
```
Comment (rbev) [inline src/ServiceControl.Persistence/DataMigration/MigrationEngine.cs]: If the `Write` call writes `checkpointAfterBatch` as given (as described in the interface specification) then how do these extra counts that are being added on from the `result` survive a crash/reboot before the next batch commits?

File: src/ServiceControl.Persistence/DataMigration/MigrationEngine.cs [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+
+                foreach (var id in result.SkippedIds)
+                {
+                    logger.LogWarning("Skipped {SourceId} in category {CategoryId}", id, category.Id);
+                }
+
+                processedThisRun += bodySkips + result.Copied + result.Skipped + result.AlreadyPresent;
+                skippedThisRun += bodySkips + result.Skipped;
+
+                if (HaltThreshold.Exceeded(skippedThisRun, processedThisRun, options.HaltThresholdPercent, options.HaltThresholdMinimum))
```
Comment (rbev) [inline src/ServiceControl.Persistence/DataMigration/MigrationEngine.cs:141]: Halt threshold is based on percent of progress rather than percent of total, does this cause a problem if there are a small number of invalid records at the start of the migration that would otherwise be ok?

**Review by rbev — COMMENTED** [review]
Partial review - adding as comments for now.

File: docs/migration/ravendb-to-sql-migration-instructions.md (hunk 1 above)
Comment (warwickschroeder) [reply docs/migration/ravendb-to-sql-migration-instructions.md:21]: The abstraction is designed to allow for possible "any direction" migrations, but there will not be an implementation for SQL as the source or Raven as the target for this version.

Yep - I'll remove that setting as it is redundant at the moment. I did think about removing this..

File: docs/migration/ravendb-to-sql-migration-overview.md (hunk 2 above)
Comment (warwickschroeder) [reply docs/migration/ravendb-to-sql-migration-overview.md:75]: Part of the design is that there will be a bunch of source and target checks (implemented at that persistence layer) that run before migration starts. Checking the schema can be part of the SQL target checks.

File: src/ServiceControl.UnitTests/Migration/MigrationEngineCategorySelectionTests.cs [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+    static MigrationEngine BuildEngine(IReadOnlyCollection<string> selectedOptionalIds, out InMemoryMigrationCheckpointStore checkpointStore)
+    {
+        checkpointStore = new InMemoryMigrationCheckpointStore();
+        var target = new InMemoryMigrationTarget(checkpointStore);
+        var options = new MigrationEngineOptions(TimeSpan.FromSeconds(1), 5, 100, selectedOptionalIds);
+        return new MigrationEngine(new InMemoryMigrationSource(), target, checkpointStore, new FakeTimeProvider(), options, NullLogger<MigrationEngine>.Instance);
+    }
+
+    [Test]
+    public void All_twelve_required_categories_are_always_selected()
```
Comment (rbev) [inline src/ServiceControl.UnitTests/Migration/MigrationEngineCategorySelectionTests.cs]: this should be an approval test so that ordering can be validated

File: src/ServiceControl.UnitTests/Migration/MigrationEngineCategorySelectionTests.cs [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+    {
+        var engine = BuildEngine([], out _);
+
+        Assert.That(engine.SelectCategories(MigrationCategoryKind.Required), Has.Count.EqualTo(12));
+    }
+
+    [Test]
+    public void Only_configured_optional_categories_are_selected()
+    {
+        var engine = BuildEngine(["EventLog"], out _);
```
Comment (rbev) [inline src/ServiceControl.UnitTests/Migration/MigrationEngineCategorySelectionTests.cs]: This should also be an approval test

File: src/ServiceControl.Persistence/DataMigration/IMigrationTarget.cs (hunk 5 above)
Comment (warwickschroeder) [reply src/ServiceControl.Persistence/DataMigration/IMigrationTarget.cs:11]: Have update the comment for it. Its a ceiling for SQL server, but more of a target for postgres. The consumer never sees this anyway

File: docs/migration/ravendb-to-sql-migration-instructions.md [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+
+| Setting | Environment variable | What it is |
+| --- | --- | --- |
+| `ServiceControl/RavenDB/ConnectionString` | `SERVICECONTROL_RAVENDB_CONNECTIONSTRING` | An external RavenDB server. Leave unset for an embedded database |
+| `ServiceControl/DbPath` | `SERVICECONTROL_DBPATH` | The embedded database's data directory |
+| `ServiceControl/RavenDB/DatabaseName` | `SERVICECONTROL_RAVENDB_DATABASENAME` | The primary database, `primary` by default |
+| `LicensingComponent/RavenDB/ThroughputDatabaseName` | `LICENSINGCOMPONENT_RAVENDB_THROUGHPUTDATABASENAME` | The throughput database, `throughput` by default |
+| `ServiceControl/RavenDB/ClientCertificatePath` or `ServiceControl/RavenDB/ClientCertificateBase64`, with `ServiceControl/RavenDB/ClientCertificatePassword` | `SERVICECONTROL_RAVENDB_CLIENTCERTIFICATEPATH` and so on | A secured external server's client certificate |
+| `ServiceControl/ErrorRetentionPeriod` | `SERVICECONTROL_ERRORRETENTIONPERIOD` | Required. Don't change it during the move |
```
Comment (johnsimons) [inline docs/migration/ravendb-to-sql-migration-instructions.md:19]: So are these exclusively being used for the migration? And if so, should then be something like `ServiceControl/Migration\....`?

File: docs/migration/ravendb-to-sql-migration-overview.md [hunk 12]
```diff
@@ -0,0 +1,294 @@
...
+- Failed error imports, the record of errors that could not be ingested
+- Group comments, **copied last of everything**, after archived and resolved messages. A comment survives only once the failed messages its group is built from have arrived, so on a large archive the comments are the last thing to appear. An empty comment field partway through a migration is the copy still running, not data loss
+- Failed message edits
+
+### Not migrated
+
+- The fifteen RavenDB index definitions, which map to a much smaller set of ordinary SQL indexes, and two of which are dead already
+- The transient in-flight collections, which are empty when nothing is running: `RetryBatches`, `RetryBatchNowForwardings`, `FailedMessageRetries`, `ArchiveOperations` and `UnarchiveOperations`
+- `ArchiveBatches` and `UnarchiveBatches`, which exist only because of how RavenDB works
+- `ConnectedApplications`, which only versions 6.0 and 6.1 wrote and nothing has read since
```
Comment (johnsimons) [inline docs/migration/ravendb-to-sql-migration-overview.md:178]: This is used for when the user is using the MassTransitConnector and this data is passed to ServicePulse to toggle features on/off

File: docs/migration/ravendb-to-sql-migration-overview.md [hunk 13]
```diff
@@ -0,0 +1,294 @@
...
+- Archived and resolved failed messages: the biggest category by far, and most of the copying time
+- The event log, without which the ServicePulse activity feed starts empty
+- Custom checks, which cost almost nothing to skip because every check re-reports on its next interval
+- Failed error imports, the record of errors that could not be ingested
+- Group comments, **copied last of everything**, after archived and resolved messages. A comment survives only once the failed messages its group is built from have arrived, so on a large archive the comments are the last thing to appear. An empty comment field partway through a migration is the copy still running, not data loss
+- Failed message edits
+
+### Not migrated
+
+- The fifteen RavenDB index definitions, which map to a much smaller set of ordinary SQL indexes, and two of which are dead already
```
Comment (johnsimons) [inline docs/migration/ravendb-to-sql-migration-overview.md]: pointless mentioning?

File: docs/migration/ravendb-to-sql-migration-overview.md [hunk 14]
```diff
@@ -0,0 +1,294 @@
...
+- The transient in-flight collections, which are empty when nothing is running: `RetryBatches`, `RetryBatchNowForwardings`, `FailedMessageRetries`, `ArchiveOperations` and `UnarchiveOperations`
+- `ArchiveBatches` and `UnarchiveBatches`, which exist only because of how RavenDB works
+- `ConnectedApplications`, which only versions 6.0 and 6.1 wrote and nothing has read since
+- Integration events still waiting to be sent when you switch over are never sent
```
Comment (johnsimons) [inline docs/migration/ravendb-to-sql-migration-overview.md:179]: anyway to mitigate this?

File: docs/migration/ravendb-to-sql-migration-overview.md [hunk 15]
```diff
@@ -0,0 +1,294 @@
...
+- Throughput history
+- Retry operations, unacknowledged and historic. One category, because RavenDB holds both lists in a single document
+- Licensing report masks
+- The uploaded licensed endpoint details file, which nothing recomputes: skipping it means the customer re-downloads it from the licence portal and uploads it again
+- Subscriptions
+
+### Optional
+
+- Archived and resolved failed messages: the biggest category by far, and most of the copying time
+- The event log, without which the ServicePulse activity feed starts empty
```
Comment (johnsimons) [inline docs/migration/ravendb-to-sql-migration-overview.md:167]: I think I still see beneficial of bringing in at least the last 7 days or something like that

File: docs/migration/ravendb-to-sql-migration-overview.md [hunk 16]
```diff
@@ -0,0 +1,294 @@
...
+- Custom checks, which cost almost nothing to skip because every check re-reports on its next interval
```
Comment (johnsimons) [inline docs/migration/ravendb-to-sql-migration-overview.md:168]: is this related to heartbeats?
I am asking becasue there is no mention of those?
One thing to watch out, is an instance is down, we will not get a heartbeat so we want to ensure that we somehow assume that everything is ok.

File: docs/migration/ravendb-to-sql-migration-overview.md [hunk 17]
```diff
@@ -0,0 +1,294 @@
...
+- Failed error imports, the record of errors that could not be ingested
```
Comment (johnsimons) [inline docs/migration/ravendb-to-sql-migration-overview.md:169]: I don't know is we can ignore these at all, at the end of the day they are error messages that could not be ingested, we may need to notify the customer about them!

File: docs/migration/ravendb-to-sql-migration-overview.md [hunk 18]
```diff
@@ -0,0 +1,294 @@
...
+- Group comments, **copied last of everything**, after archived and resolved messages. A comment survives only once the failed messages its group is built from have arrived, so on a large archive the comments are the last thing to appear. An empty comment field partway through a migration is the copy still running, not data loss
```
Comment (johnsimons) [inline docs/migration/ravendb-to-sql-migration-overview.md:170]: This is important to bring in, as soon as we create the group, becasue there may be info there about informing things like "do not retry this group...."

File: docs/migration/ravendb-to-sql-migration-overview.md [hunk 19]
```diff
@@ -0,0 +1,294 @@
...
+- Failed message edits
```
Comment (johnsimons) [inline docs/migration/ravendb-to-sql-migration-overview.md:171]: What is this?

File: docs/migration/ravendb-to-sql-migration-overview.md (hunk 3 above)
Comment (warwickschroeder) [reply docs/migration/ravendb-to-sql-migration-overview.md]: Throughput history is in the required set of categories, so it will be migrated before SC starts up. This means that when the migration starts, the throughput table in SQL is empty so nothing to merge with. Set rather than add makes the migration resumable and the checkpoint stops a second pass overwriting days the collectors have written since.

I've update the overview doc toclarify this.

File: docs/migration/ravendb-to-sql-migration-overview.md (hunk 4 above)
Comment (warwickschroeder) [reply docs/migration/ravendb-to-sql-migration-overview.md]: Looking into what exactly raven sets to expire, its only resolved or archived FailedMessages, and EventLogItems. Both arent in the required categories, so during the migration of the required set, nothing is expired from under us. If the optional records are past their retention already, they are skipped. If a row is swept away by raven mid-copy, it is an absence rather than a skip.

I'll update the overview doc to clarify this.

File: src/ServiceControl.Persistence/DataMigration/MigrationEngine.cs (hunk 7 above)
Comment (warwickschroeder) [reply src/ServiceControl.Persistence/DataMigration/MigrationEngine.cs]: Fixed in https://github.com/Particular/ServiceControl/pull/5894/changes/909e4bfeaa36503e62c1d56aaccd11d8c6ab1300#diff-14df616bb97776e63067960a909ea5a788d33570699a00737b6053d4708b9f34.

I've added a sequence diagram to the overview doc to clearly show what the migration engine does and how the checkpoint/resume functionality works.

**Review by johnsimons — APPROVED** [review]
A few question and comments but overall, lets keep going

File: src/ServiceControl.Persistence/DataMigration/MigrationEngine.cs (hunk 8 above)
Comment (warwickschroeder) [reply src/ServiceControl.Persistence/DataMigration/MigrationEngine.cs:141]: A small number cannot halt anything, because the floor gates the percentage entirely. But a cluster of more than 100 bad rows near the start does halt a category, which is probably correct.

One bug I did notice though is that legitimate skips (past retention, orphaned group comments, etc) are being counted towards a halt. These types of skips should not count towards it. Will fix.

I've also added a section to the overview doc with a diagram to illustrate this.

File: src/ServiceControl.UnitTests/Migration/MigrationEngineCategorySelectionTests.cs (hunk 9 above)
Comment (warwickschroeder) [reply src/ServiceControl.UnitTests/Migration/MigrationEngineCategorySelectionTests.cs]: done

File: src/ServiceControl.UnitTests/Migration/MigrationEngineCategorySelectionTests.cs (hunk 10 above)
Comment (warwickschroeder) [reply src/ServiceControl.UnitTests/Migration/MigrationEngineCategorySelectionTests.cs]: done

File: src/ServiceControl.Persistence/DataMigration/MigrationEngine.cs (hunk 8 above)
Comment (warwickschroeder) [reply src/ServiceControl.Persistence/DataMigration/MigrationEngine.cs:141]: Have also added some more tests to test all paths of the halting logic

File: src/ServiceControl.Persistence/DataMigration/MigrationEngine.cs (hunk 7 above)
Comment (warwickschroeder) [reply src/ServiceControl.Persistence/DataMigration/MigrationEngine.cs]: Have added more tests for all paths and edge cases of the checkpoint/resume logic

File: docs/migration/ravendb-to-sql-migration-overview.md (hunk 17 above)
Comment (rbev) [reply docs/migration/ravendb-to-sql-migration-overview.md:169]: Probably worth just marking them as errors in the pre-flight and force them to resolve before migrate!

File: docs/migration/ravendb-to-sql-migration-instructions.md (hunk 11 above)
Comment (warwickschroeder) [reply docs/migration/ravendb-to-sql-migration-instructions.md:19]: These just list the current settings for raven, which the migration engine uses to set its "Source", but yes, any new setting thats migration specific will be "ServiceControl/Migration/xxxxx"

File: docs/migration/ravendb-to-sql-migration-overview.md (hunk 19 above)
Comment (warwickschroeder) [reply docs/migration/ravendb-to-sql-migration-overview.md:171]: https://github.com/Particular/ServiceControl/blob/7dd5105f295f81f3db23d8b0a44dec28e96d072f/src/ServiceControl.Persistence.RavenDB/Editing/EditFailedMessagesDataStore.cs#L9

I've expanded the dot point to add more clarity.
